### PR TITLE
Disallow trailing data in value.FromJSON and fieldpath.FromJSON and add streaming token tests

### LINKED ISSUE #318: Validate Token Type Errors in readObject
### Background
During the code review for the `encoding/json/v2` migration in PR [#292](https://github.com/kubernetes-sigs/structured-merge-diff/pull/292), it was identified that we need stricter validation around token types during object parsing.

### Description
We need to develop a test case to explicitly confirm that the parser correctly handles unexpected token types. Specifically, if the parser encounters a token type other than `EndObject` or `String` at a position where a JSON key is expected, the system must return the appropriate, well-defined error. 

This ensures that malformed JSON is caught early and produces a standard error matching previous behavior.

### Code Pointers
* **File:** [`typed/parser.go`](https://github.com/kubernetes-sigs/structured-merge-diff/pull/292/files#diff-e55e5828452f14643c75edc316f43e06f9e5b1580f4f9104b2b2361623194a2f)
* **Function to Test:** `readObject`
* **Reference PR Comment:** [Discussion Link](https://github.com/kubernetes-sigs/structured-merge-diff/pull/292#discussion_r2589587228)

### Metadata
* **Priority:** P0
* **Complexity:** Medium

### PR DESCRIPTION
**Pull request by jpbetz**
Fixes https://github.com/kubernetes-sigs/structured-merge-diff/issues/318

### CODE REVIEW

File: fieldpath/serialize_test.go [hunk 1]
```diff
@@ -313,6 +313,60 @@ func TestDeserializeForValidNonNormalized(t *testing.T) {
...
+
+		{"keyEmptyString", `{"":{}}`},
+		{"keyMissingColon", `{"x":{}}`},
+		{"keyOnlyColon", `{":":{}}`},
+
+		// TODO: Trailing data should not be allowed
+		//{"trailingClosed", `{"f:a":{}}}`},
+		//{"trailingObject", `{"f:a":{}}{}`},
+		//{"trailingValue", `{"f:a":{}}"string"`},
+	}
```
Comment (jpbetz) [inline fieldpath/serialize_test.go]: Looks like jsonIter consumes the entire buffer, but our token parsing ignores trailing data.

Patch now?  My understanding is that jsonv2 is strict on this.

cc @liggitt @lalitc375

File: fieldpath/serialize_test.go [hunk 2]
```diff
@@ -313,6 +313,60 @@ func TestDeserializeForValidNonNormalized(t *testing.T) {
...
+		{"empty", ``},
+		{"whitespaceOnly", `   `},
+
+		{"numberAsKey", `{1:{}}`},
+		{"trailingCommaAfterEntry", `{"f:a":{},}`},
+		{"unclosedAfterOpen", `{`},
+		{"unclosedAfterValue", `{"f:a":{}`},
+		{"unclosedAfterTrailingComma", `{"f:a":{},`},
+
+		{"closedBeforeComma", `{"f:a" }:{}}`},
```
Comment (liggitt) [inline fieldpath/serialize_test.go]: not sure I understand the "closedBeforeComma" name for this one

File: fieldpath/serialize_test.go [hunk 3]
```diff
@@ -313,6 +313,60 @@ func TestDeserializeForValidNonNormalized(t *testing.T) {
...
+		{"keyWithoutColon", `{"f:a"}`},
+		{"leadingCommaInsideValue", `{"f:a":{,}}`},
+
+		{"bareCloseBrace", `}`},
+
+		{"keyEmptyString", `{"":{}}`},
+		{"keyMissingColon", `{"x":{}}`},
+		{"keyOnlyColon", `{":":{}}`},
+
+		// TODO: Trailing data should not be allowed
```
Comment (liggitt) [inline fieldpath/serialize_test.go]: also make sure we have test cases capturing whether internal and trailing whitespace is allowed or treated as an error (if we don't already have that ... we may)

File: fieldpath/serialize_test.go (hunk 1 above)
Comment (liggitt) [reply fieldpath/serialize_test.go]: Yeah, it seems wrong to consume the whole buffer but ignore semantically invalid trailing tokens. Tightening that up to make sure the remaining buffer is either empty or only contains whitespace sounds good to me.

I have a hard time telling if there's any actual exposure from the top-level managedFields or not. Before fixing this, might be worth a sanity check test demonstrating whatever our entry-points are for parsing managedFields would fail if a field path tried to set one of these invalid trailing data things.

File: fieldpath/serialize_test.go (hunk 3 above)
Comment (jpbetz) [reply fieldpath/serialize_test.go]: Very good catch.

File: fieldpath/serialize_test.go (hunk 2 above)
Comment (jpbetz) [reply fieldpath/serialize_test.go]: I think I copy-pasted a test case when adding one and forgot to rename it.  Fixing.

**Comment by jpbetz** [comment]
> Yeah, it seems wrong to consume the whole buffer but ignore semantically invalid trailing tokens. Tightening that up to make sure the remaining buffer is either empty or only contains whitespace sounds good to me.

Added.

> I have a hard time telling if there's any actual exposure from the top-level managedFields or not. Before fixing this, might be worth a sanity check test demonstrating whatever our entry-points are for parsing managedFields would fail if a field path tried to set one of these invalid trailing data things.

I've looked into this. I cannot find any exposure. For Protobuf/CBOR all the data is famed, so if a writer adds trailing data, the reader just ignores it.  For JSON, [FIeldV1](https://github.com/kubernetes/kubernetes/blob/e8742a6ba73fba3729ed9b44b89dd9540f48fd5f/staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/fieldsv1.go#L89) handles the marshallening and there is no opportunity to smuggle bytes into the surrounding metadata fields.

**Comment by jpbetz** [comment]
Fix is in place for trailing and tests are added.

File: value/value_test.go [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+
+		{"bareCloseBrace", `}`},
+		{"bareCloseBracket", `]`},
+		{"bareComma", `,`},
+		{"bareColon", `:`},
+
+		// TODO: Trailing data should not be allowed
+		//{"trailingClosed", `{"f:a":{}}}`},
+		//{"trailingObject", `{"f:a":{}}{}`},
+		//{"trailingValue", `{"f:a":{}}"string"`},
```
Comment (liggitt) [inline value/value_test.go]: should we be able to uncomment these and they error now?

Comment (jpbetz) [reply value/value_test.go]: Fixed

**Comment by jpbetz** [comment]
@liggitt I think this one is ready

**Comment by liggitt** [comment]
lgtm, before merge can you open a dry-run bump to your branch in k/k to make sure everything there is happy with this as well?

**Comment by jpbetz** [comment]
Dry run test: https://github.com/kubernetes/kubernetes/pull/139000

**Comment by liggitt** [comment]
dry run was green

/hold cancel
